src/engagement_db_to_analysis/code_imputation_functions.py

```diff
@@ -134,26 +150,134 @@ def _impute_age_category(user, messages_traced_data, analysis_dataset_configs):
+
+            elif coding_config.kenya_analysis_location == KenyaAnalysisLocations.CONSTITUENCY:
+                log.info(f"Found kenya_analysis_location in {coding_config.analysis_dataset} coding config")
+
+                assert constituency_coding_config is None, f"Found more than one constituency_coding_config in " \
+                    f"analysis_dataset_config"
+                constituency_coding_config = coding_config
+                location_engagement_db_datasets = analysis_dataset_config.engagement_db_datasets
+
+            elif coding_config.kenya_analysis_location == KenyaAnalysisLocations.COUNTY:
+                log.info(f"Found kenya_analysis_location in {coding_config.analysis_dataset} coding config")
+
+                assert county_coding_config is None, f"Found more than one county_coding_config in " \
+                    f"analysis_dataset_config"
+                county_coding_config = coding_config
+
+    assert county_coding_config is not None, f"No county_coding_config in " \
+        f"analysis_dataset_config"
+    assert constituency_coding_config is not None, f"No constituency_coding_config in " \
+        f"analysis_dataset_config"
+
+    for message_traced_data in messages_traced_data:
+        if message_traced_data["dataset"] in location_engagement_db_datasets:
+            message = Message.from_dict(dict(message_traced_data))
+            message_analysis_config = analysis_dataset_config_for_message(analysis_dataset_configs, message)
+
+            # Up to 1 location code should have been assigned in Coda. Search for that code,
+            # ensuring that only 1 has been assigned or, if multiple have been assigned, that they are non-conflicting
+            # control codes
+            location_code = None
+            for coding_config in message_analysis_config.coding_configs:
+                latest_coding_config_labels = get_latest_labels_with_code_scheme(message, coding_config.code_scheme)
+
+                if len(latest_coding_config_labels) > 0:
+                    latest_coding_config_label = latest_coding_config_labels[0]
+                    if latest_coding_config_label.checked:
+                        coda_code = coding_config.code_scheme.get_code_with_code_id(latest_coding_config_label.code_id)
+                        if location_code is not None:
+                            if location_code.code_id != coda_code.code_id:
+                                location_code = constituency_coding_config.code_scheme.get_code_with_control_code(Codes.CODING_ERROR)
+                        else:
+                            location_code = coda_code
+
+            # If no code was found, then this location is still not reviewed.
+            # Synthesise a NOT_REVIEWED code accordingly.
+            if location_code is None:
+                location_code = constituency_coding_config.code_scheme.get_code_with_control_code(Codes.NOT_REVIEWED)
```

Comment:
Is this necessary?

Reply:
If we _impute_not_reviewed_labels before imputing kenyan-location-labels this is not necessary. Removed.